### ISSUE TITLE
False negative: Spotbugs should complain when storing returned Nullable into NonNull variable

### ISSUE THREAD

**Issue by ghernadi** [spotbugs/spotbugs#3113]
A quite simple reproducer:

```
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import java.util.Random;

public class NonNullLocalVariable
{
    private static final Random RANDOM = new Random();

    public static void main(String[] args)
    {
        @Nonnull Object test = random();
        System.out.println(test); // only to suppress "not used" warning
    }

    private static @Nullable Object random()
    {
        return RANDOM.nextBoolean() ? null : "asdf";
    }
}
```
I have not created a proper unit test for this simply because I do not know what bug-code to expect here.

**Comment by gtoison** [comment]
Thanks for reporting the issue. I think that there are two problems here:
- SpotBugs does not seem to check for local variables annotated with `@Nonnull`, don't quote me on this because I'd have to check
- `@Nullable` has a misleading name (IMO) and actually means that nullness is unknown, as the javadoc says:

>  This annotation is useful mostly for overriding a `Nonnull` annotation. Static analysis tools should generally treat the annotated items as though they had no annotation, unless they are configured to minimize false negatives. Use `@CheckForNull` to indicate that the element value should always be checked for a null value.

So you should be using `@CheckForNull` here instead of `@Nullable`

**Comment by ghernadi** [comment]
Do you want to keep this issue open regarding your first point? If not, I'm happy if we close this issue.

**Comment by gtoison** [comment]
I quickly tested and using `@CheckForNull` seems to resolve the problem.
The potentially misleading name of `@Nullable` and its semantic are here to stay, so I think you can close this issue.
